monkey/infection_monkey/network/windowsserver_fingerprint.py

```diff
@@ -0,0 +1,116 @@
+"""
+Implementation from https://github.com/SecuraBV/CVE-2020-1472
+"""
+
+import logging
+import subprocess
+
+from impacket.dcerpc.v5 import epm, nrpc, transport
+
+import infection_monkey.config
+from infection_monkey.network.HostFinger import HostFinger
+from infection_monkey.utils.environment import is_windows_os
+
+LOG = logging.getLogger(__name__)
+
+
+class WindowsServerFinger(HostFinger):
+    # Class related consts
+    MAX_ATTEMPTS = 2000
+    _SCANNED_SERVICE = "Windows Server"
+
+    def __init__(self):
+        self._config = infection_monkey.config.WormConfiguration
+
+    def get_dc_name(self, DC_IP):
+        """
+        Gets NetBIOS name of the DC.
+        """
+        name = ''
+        try:
+            if is_windows_os():
+                cmd = f'nbtstat -A {DC_IP} | findstr "<00>"'
+                name = subprocess.check_output(cmd, shell=True).decode().split('\n')[0].strip(' ').split(' ')[0]
+            else:
+                cmd = f'nmblookup -A {DC_IP} | grep "<00>"'
+                name = subprocess.check_output(cmd, shell=True).decode().split('\n')[0].strip('\t').strip(' ').split(' ')[0]
+        except BaseException as ex:
+            LOG.info(f'Exception: {ex} Most likely not a Windows DC.')
+        return name
+
+    def get_host_fingerprint(self, host):
+        """
+        Checks if the Windows Server is vulnerable to Zerologon.
+        """
+        unexpected_error_encountered = False
+
+        def try_zero_authenticate(DC_HANDLE, DC_IP, DC_NAME):
```

Comment:
Why is this method defined in a method?

Reply:
It's being [used in a loop](https://github.com/guardicore/monkey/pull/868/files/7bdc7ce4c2690e48dd42499226caeda1cbfebb7b#diff-43d174fcf319ecc99790cc783c435aa05c3ae654a7b0c60fe47c217ac5c4dbb1R99) and it's only relevant inside the `get_host_fingerprint` method.

Reply:
My suggestion would be to not to use methods in methods when possible. If you ever need to reuse it somewhere else, you'll have to extract it. But the main problem is readability - you start reading one method and immediately are introduced to another. Ideal solution would be to have `try_zero_authenicate` after `get_host_fingerprint`.